pkg/member/participant.go

```diff
@@ -0,0 +1,405 @@
+	return m.leadership.Campaign(leaseTimeout, m.MemberValue())
+}
+
+// KeepLeader is used to keep the PD leader's leadership.
+func (m *Participant) KeepLeader(ctx context.Context) {
+	m.leadership.Keep(ctx)
+}
+
+// CheckLeaderPrecheck does some pre-check before checking whether or not it's the leader.
+// It returns true if it passes the pre-check, false otherwise.
+func (m *Participant) CheckLeaderPrecheck() error {
+	// No specific thing to check. Returns no error.
+	return nil
+}
+
+// CheckLeader checks returns true if it is needed to check later.
+func (m *Participant) CheckLeader() (*pdpb.Member, int64, bool) {
+	if err := m.CheckLeaderPrecheck(); err != nil {
+		log.Error("failed to pass pre-check, check the leader later", errs.ZapError(errs.ErrEtcdLeaderNotFound))
+		time.Sleep(200 * time.Millisecond)
+		return nil, 0, true
+	}
+
+	leader, rev, err := election.GetLeader(m.client, m.GetLeaderPath())
+	if err != nil {
+		log.Error("getting the leader meets error", errs.ZapError(err))
+		time.Sleep(200 * time.Millisecond)
+		return nil, 0, true
+	}
+	if leader != nil {
+		if m.isSameLeader(leader) {
+			// oh, we are already the leader, which indicates we may meet something wrong
+			// in previous CampaignLeader. We should delete the leadership and campaign again.
+			log.Warn("the leader has not changed, delete and campaign again", zap.Stringer("old-leader", leader))
+			// Delete the leader itself and let others start a new election again.
+			if err = m.leadership.DeleteLeaderKey(); err != nil {
+				log.Error("deleting the leader key meets error", errs.ZapError(err))
+				time.Sleep(200 * time.Millisecond)
+				return nil, 0, true
+			}
+			// Return nil and false to make sure the campaign will start immediately.
+			return nil, 0, false
+		}
+	}
+	return leader, rev, false
+}
+
+// WatchLeader is used to watch the changes of the leader.
+func (m *Participant) WatchLeader(serverCtx context.Context, leader *pdpb.Member, revision int64) {
+	m.setLeader(leader)
+	m.leadership.Watch(serverCtx, revision)
+	m.unsetLeader()
+}
+
+// ResetLeader is used to reset the PD member's current leadership.
+// Basically it will reset the leader lease and unset leader info.
+func (m *Participant) ResetLeader() {
+	m.leadership.Reset()
+	m.unsetLeader()
+}
+
+// isSameLeader checks whether a server is the leader itself.
+func (m *Participant) isSameLeader(leader *pdpb.Member) bool {
+	return leader.GetMemberId() == m.ID()
+}
+
+// CheckPriority checks whether there is another participant has higher priority and resign it as the leader if so.
+func (m *Participant) CheckPriority(ctx context.Context) {
+	// TODO: implement weighted-election when it's in need
+}
+
+// InitMemberInfo initializes the member info.
+func (m *Participant) InitMemberInfo(advertiseClientUrls, advertisePeerUrls, name string, rootPath string) {
+	leader := &pdpb.Member{
+		Name:       name,
+		MemberId:   m.ID(),
+		ClientUrls: strings.Split(advertiseClientUrls, ","),
```

Comment:
Why do we have these fields?

Reply:
removed ClientUrls and PeerUrls. We reuse pdpb.Member for now with the following TODO so that we can start integration first.

	// TODO: use a more general instead of pdpb.Member. The latter is tightly
	// coupled with the embedded etcd. For now, we can reuse it and ignore the
	// fields which are related to embedded etcd, such as PeerUrls and ClientUrls.